python/torch_mlir/extras/fx_importer.py

```diff
@@ -1485,6 +1585,74 @@ def _import_torch_op_overload(
         for i, value in enumerate(operation.results):
             self.bind_node_value(node, value, i)
 
+    def _import_shape_symbols_with_guards(
+        self, loc: Location, symbolic_guards: Dict[str, RangeConstraint]
+    ):
+        for symbol, constraints in symbolic_guards.items():
+            # Create torch.sym_int ops
+            operation = Operation.create(
+                name="torch.symbolic_int",
+                attributes={
+                    "symbol_name": StringAttr.get(symbol),
+                    "min_val": self._cc.integer_attr(constraints.min_val, 64),
+                    "max_val": self._cc.integer_attr(constraints.max_val, 64),
+                },
+                results=[self._cc.torch_int_type],
+                loc=loc,
+            )
+            self.bind_symbol_value(symbol, operation.result)
+
+    def _create_bind_symbolic_shape_ops(self, loc: Location, node: torch_fx.Node):
+        node_val = node.meta.get("val")
+        if (node_val is not None) and isinstance(node_val, TorchFakeTensor):
+            # Only create bind ops if the shapes contain symbolic sizes
+            if node_val._has_symbolic_sizes_strides:
+                # Read node metadata to obtain shape symbols and expressions
+                symbols_set = set()
+                shape_exprs = []
+                for s in node_val.size():
+                    if isinstance(s, torch.SymInt):
+                        symbols_set.update(s.node.expr.free_symbols)
+                        shape_exprs.append(s.node.expr)
+                    else:
+                        assert isinstance(s, int)
+                        shape_exprs.append(s)
+
+                # Map from actual shape symbols to local symbols in the affine map
+                symbols_set = sorted(symbols_set, key=lambda x: x.name)
+
+                symbol_table = {
+                    str(symbol): AffineSymbolExpr.get(i)
+                    for i, symbol in enumerate(symbols_set)
+                }
+
+                # Convert symbolic shape expressions into affine expressions
+                affine_exprs = [
+                    (
+                        AffineConstantExpr.get(expr)
+                        if isinstance(expr, int)
+                        # Use `symbol_table.copy()` here to avoid `eval` from
+                        # adding the `__builtins__` entry to the map.
+                        else eval(str(expr), symbol_table.copy())
+                    )
+                    for expr in shape_exprs
+                ]
+                affine_map = AffineMap.get(0, len(symbols_set), affine_exprs)
+
+                # Build operand list
+                operand_list = []
+                operand_list.append(self.resolve_node_value(node))
+                for symbol in symbol_table.keys():
+                    operand_list.append(self.resolve_symbol_value(symbol))
+
+                # Create torch.bind_symbolic_shape ops
+                Operation.create(
```

Comment:
I think we need to make the emission of these ops optional/configurable, even if just as a transitional measure while phased in: without a cleanup pass wired into every pipeline, these impure ops will persist and block a number of optimizations if added. Probably best to add a knob to FxImporter to enable symbolic shape guards and then gate anything that calls this function on it.

Reply:
Just noticed you explicitly mentioned gating emission of the impure ops, but I've gated the insertion of both the pure `symbolic_int` ops as well as these impure `bind_symbolic_shape_ops`, just to provide a way to completely disable this feature if needed (we can relax this once this becomes a bit more load bearing).